### PR TITLE
Add docs including automated deployment

### PR DESCRIPTION
**Pull request by blazekadam**

### CODE REVIEW

File: docs/tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+     - LogVariables
+     - SaveBest
+
+Finally run the training with:
+
+.. code-block:: bash
+
+   cxflow train config.yaml
+
+A fully functional example may be found at TODO link.
```
Comment (petrbel) [inline docs/tutorial.rst]: TODO link

File: docs/tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+to the ``_create_model`` function. This way, the whole model can be easily 
+parametrized.
+
+.. tip::
+   Try to experiment with the ``fc_size`` parameter. How small the fully connected layer can be before the performance
+   degrades?
+
+Next steps
+----------
+See TODO for additional examples or read the :py:class:`cxflow_tensorflow.BaseModel` reference
```
Comment (petrbel) [inline docs/tutorial.rst]: todo

File: docs/multi_gpu.rst [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+.. code-block:: yaml
+   :caption: config.yaml
+   :emphasize-lines: 3
+
+   model:
+      class: nets.MyMultiGPUNet
+      n_gpus: 4
+      ...
+
+That's it! Now better check your server temperatures...
```
Comment (petrbel) [inline docs/multi_gpu.rst]: I don't like this, it's way too informal... I mean, this shoudn't be a paper-level English, right, but I feel it's too much.

**Comment by petrbel** [comment]
I added some comments to few lines.

**Comment by blazekadam** [comment]
We need to merge this ASAP. I suggest to 
- fix the minor Issues proposed by Petr
- merge the other PRs
- merge dev to doc
- merge doc to dev and continue to update it via PRs

File: docs/multi_gpu.rst (hunk 3 above)
Comment (blazekadam) [reply docs/multi_gpu.rst]: I agree. I ll update it.

File: docs/index.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+- Issue Tracker: https://github.com/Cognexa/cxflow-tensorflow/issues
+- Source Code: https://github.com/Cognexa/cxflow-tensorflow
+
+Support
+-------
+
+If you are having issues, please let us know either via the official
+`issue tracker <https://github.com/Cognexa/cxflow-tensorflow/issues>`_ or via e-mail.
+We have a mailing list located at: cxflow@cognexa.com
```
Comment (FloopCZ) [inline docs/index.rst]: It is not a mailing list, is it? I would write just:
via e-mail on cxflow@cognexa.com.

File: docs/multi_gpu.rst [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+   :emphasize-lines: 3
+
+   model:
+      class: nets.MyMultiGPUNet
+      n_gpus: 4
+      ...
+
+That's it! Now run the **cxflow** training and check if all your GPUs are in fact utilized.
+
+A fully working example may be found HERE.
```
Comment (FloopCZ) [inline docs/multi_gpu.rst]: fix HERE

File: docs/multi_gpu.rst [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+   Tower architecture for multi GPU training.
+
+In order to leverage the true potential of multiple GPUs, **cxflow-tensorflow** creates a bit more complicated
+computational graph than usual. In fact, the whole model is multiplicated so that each GPU has its own *tower*. The
+variables to be trained are, naturally, shared between the towers. Each GPU computes its own feed-forward pass and the
+respective variable gradients. Consequently, the gradients are averaged and used for update.
+
+During the training **cxflow-tensorflow** distributes each batch equally among the towers. The fetched outputs are
+concatenated preserving its orders which ensures perfect abstraction for the remaining **cxflow** components. From the
```
Comment (FloopCZ) [inline docs/multi_gpu.rst]: its orders -> their order

File: docs/multi_gpu.rst [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+perspective of hooks and datasets, nothing has changed. Additionally, **cxflow-tensorflow** handles
+incomplete batches with ease.
```
Comment (FloopCZ) [inline docs/multi_gpu.rst]: with ease -> automatically
so that it does not sound like you have to do something for it

File: docs/tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+########
+
+Imagine you colleague prepared a **cxflow** compatible dataset for the well 
+known task of recognizing images of hand-written letters. Now, your task is to 
+implement a baseline neural network for that. It can not be done easier than
+with cxflow-tensorflow.
+
+First cxflow-tensorflow model
+-----------------------------
+A simple convolution network takes only a couple of lines.
```
Comment (FloopCZ) [inline docs/tutorial.rst]: convolutional

File: docs/tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+.. code-block:: python
+  :caption: convnet.py
+
+   import tensorflow as tf
+   import tensorflow.contrib.slim as slim
+
+   from cxflow_tensorflow import BaseModel
```
Comment (FloopCZ) [inline docs/tutorial.rst]: I suggest to use namespaces in all our tutorials (it is generally a good practice).
`import cxflow_tensorflow as cxtf`

File: docs/tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+   favorite framework such as `Keras <https://keras.io/>`_,
+   `Slim <https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/slim>`_ or vanilla TensorFlow.
+
+The next step is to write a simple configuration file putting everything together:
+
+.. code-block:: yaml
+   :caption: config.yaml
+
+   dataset:
+      class datasets.MNISTDataset
```
Comment (FloopCZ) [inline docs/tutorial.rst]: missing `:` after class

File: docs/tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+      class: convnet.SimpleConvNet
+      optimizer:
+         class: AdamOptimizer
+         learning_rate: 0.001
+      inputs: [images, labels]
+      outputs: [loss, probs]
+   hooks:
+     - ComputeStats:
+         variables:
+           'loss':[mean]
```
Comment (FloopCZ) [inline docs/tutorial.rst]: missing space, extra quotes

File: docs/tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+         learning_rate: 0.001
+      inputs: [images, labels]
+
+Model parameters
+~~~~~~~~~~~~~~~~
+Note that the model (hyper-)parameters such as the number of layers were all hard-coded in our example.
+Contrary to that, those parameters happen to frequently change as we search for the best performing configuration.
+
+In **cxflow**, model parameters may be defined and configured quite easily.
+For example, to introduce new ``fc_size`` parameter controlling the number of neurons in the fully connected,
```
Comment (FloopCZ) [inline docs/tutorial.rst]: fully connected layer

File: docs/tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+   Try to experiment with the ``fc_size`` parameter. How small the fully connected layer can be before the performance
+   degrades?
+
+Next steps
+----------
+See our `GitHub examples repository <https://github.com/Cognexa/cxflow-examples>`_.
+for additional examples or read the :py:class:`cxflow_tensorflow.BaseModel` reference for the full list of
+customization options.
+
+This project contains additional util functions and **cxflow** hooks documented in the:doc:`cxflow_tensorflow/index`.
```
Comment (FloopCZ) [inline docs/tutorial.rst]: missing space between "the" and :doc:

File: .circleci/config.yml [hunk 14]
```diff
@@ -34,8 +34,21 @@ references:
       name: PyPI deploy
       command: |
         pip3 install wheel setuptools --upgrade
-        ./.circle_pypi_deploy.sh
+        .circleci/pypi_deploy.sh
```
Comment (FloopCZ) [inline .circleci/config.yml]: this should be sourced from the [ci-utils](https://github.com/Cognexa/ci-utils) repo (master is @petrbel)

File: .circleci/config.yml [hunk 15]
```diff
@@ -34,8 +34,21 @@ references:
...
+  add_rw_key: &add_rw_key
+    add_ssh_keys:
+      fingerprints:
+        - "04:f1:4a:89:30:eb:e6:d0:d7:e5:0b:fc:f4:72:12:d8"
+
+  doc: &doc
+    run:
+      name: Generate documentation
+      command: |
+        .circleci/doc_deploy.sh
```
Comment (FloopCZ) [inline .circleci/config.yml]: ditto

File: docs/_static/highlight.css [hunk 16]
```diff
@@ -0,0 +1,75 @@
+dl.class, dl.function{
```
[the remark this answers is not in the document]
Comment (blazekadam) [reply docs/_static/highlight.css]: It was all written by hand :)

This should be moved to a shared repo anyways...

File: docs/conf.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+                       'cxflow': ('https://cxflow.org', None)}
+
+
+def setup(app):
+    app.add_stylesheet("highlight.css")
+
+import os
+import sys
+sys.path.insert(0, os.path.abspath('..'))
+sys.path.insert(0, os.path.abspath('../..'))
```
[the remark this answers is not in the document]
Comment (blazekadam) [reply docs/conf.py]: ditto

File: docs/tutorial.rst (hunk 10 above)
Comment (blazekadam) [reply docs/tutorial.rst]: This example is yet to be done in cxflow-examples. The code blocks will be updated afterwards...

File: docs/tutorial.rst (hunk 9 above)
Comment (blazekadam) [reply docs/tutorial.rst]: This example is yet to be done in cxflow-examples. The code blocks will be updated afterwards...

**Comment by blazekadam** [comment]
Thanks, I fixed most of your comments, the remaining are open issues now...

**Pull request merged**